lib/src/steps/result.dart

```diff
@@ -0,0 +1,312 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:collection' show UnmodifiableListView;
+import 'dart:math' show min, max;
+
+import 'package:collection/collection.dart' show UnmodifiableSetView;
+import 'package:source_span/source_span.dart';
+import 'package:meta/meta.dart' show required, sealed;
+
+import '../model.dart' show Suggestion, SuggestionLevel;
+import 'run.dart' show run;
+import 'step.dart' show Step;
+import 'tags.dart';
+
+/// Data class for holding all analysis results.
+@sealed
+class Result {
+  final Set<String> tags;
+  final List<Suggestion> suggestions;
+
+  Result._({
+    this.tags,
+    this.suggestions,
+  });
+}
+
+@sealed
+class ResultBuilder {
+  final String _suggestionCodePrefix;
+  final List<Suggestion> _suggestions = [];
+  final Set<String> _tags = {};
+
+  ResultBuilder(this._suggestionCodePrefix);
+
+  /// Create a _suggestion_ with severity **error**.
+  ///
+  /// Severity **error** should be used when:
+  ///  * An issue _will_ casuse problems for users of the package,
+  ///  * The package is fundamentally broken,
+  ///  * The issue is to sever to warrent a [warning] or [hint].
+  ///
+  /// **Examples** include:
+  ///  * Static code errors,
+  ///  * Missing dependencies.
+  ///
+  /// A suggestion have following parameters:
+  ///  * [kind], (optional) kind identifier for statistics and grouping,
+  ///  * [title], a human-readable title for the suggestion,
+  ///  * [description], a human-readable markdown description of the suggestion,
+  ///  * [location], (optional) span of source code that caused this suggestion,
+  ///  * [penalty], (optional) penalty for package score ranging from 0 to 1.
+  void error({
+    String kind,
```

Comment:
should we make this `@required`?

Reply:
Probably... How do you feel if we give up the automatic prefixing... It's natural to write kinds like `changelog-missing` anyways...

And if we have clashing codes it's not really a big deal, anyways, we use them for stats, collapsing and testing... At least that's what I imagine..